### PR TITLE
Blake - Contact Page

### PR DESCRIPTION
**Pull request by BlakeGuilloud**
I am creating this strictly for discussion at the moment. While it all works locally, I want to confirm a decision or two before this goes into the codebase.

This PR will introduce an email service and contact form into the resources reentry hub app.

In order for this feature to be functional, we will need to introduce 5 new environmental variables into the node process. I can cut these down to two if the admins want to hardcode the AWS default region and accounts email addresses into the codebase. Just let me know.

These variables can be found in a `sample.env` file. In order to properly run this feature, you will need to create a `.env` file in your local project and supply it with proper variable declarations.

I have several questions that I will address within the code itself, but I am making this PR to allow for discussion. 

### API
Created 3 new files: 
- controller.js
  - handles incoming requests via chained service, shaper, validator, or response functions.
- service.js
  - handles business logic.
- respones.js
  - returns proper API responses to the client.

#### Email Service
We decided to use AWS Simple Email Service as our client for sending emails from the contact form. 
The `sendEmail` route utilizes this service as it accepts a payload including:
```{ name, email, message }```

The service instantiates a new SES providing appropriate AWS keys and region, builds an `email` object, and sends the request to the AWS SES service. 

I have built this out by hardcoding @ejaxon and @marshallj email addresses as the `ToAddresses`, while the email itself will come from an env variable set by one of the admins.

#### New Dependencies
- aws-sdk
  - utlizes the SES service to send emails from the app.
- body-parser
  - properly parse incoming requests from the client.
- dotenv
  - this package simply loads all variables declared in a `.env` file into the node process. Extremely useful in development environments.
- cors
  - allows cross origin

### Client
Built a simple React form that sets a component's state on input change and submits a form via the `fetch` API. Introduced a minor ternary checking for location.href including 'contact' to maintain visual consistency within the application.

### EJS
While not introduced in this PR. Using a service like EJS would be nice to format and style our emails. I figured since they were all internal I would really show off my frontend skills by setting two whole `<br />` tags to separate the emails message, email address, and sender name. Much skill. I don't have much experience using node templating services, but this could be a fun project for someone, if not me, to explore.

## Related Issue

#61 Contact Page

## How Has This Been Tested?

This has been tested in a local environment by the author using mock ToAddresses that have been authorized on AWS.

## Screenshots (if appropriate):
image
image

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/components/Contact.jsx [hunk 1]
```diff
@@ -1,21 +1,152 @@
...
+    const opts = {
+      method: 'POST',
+      body: JSON.stringify(this.state),
+      headers: new Headers({
+        'Content-Type': 'application/json',
+      }),
+    };
+
+    // this url needs to be wherever the server is getting hosted. Obviously localhost:3001 will not work on a production build.
+    const url = 'http://localhost:3001';
```
Comment (BlakeGuilloud) [inline src/components/Contact.jsx]: This is the big question I have. I need to know what to set this URL to in order to properly make an API request. An option would be to set this as a process.env variable and point it to the EC2 instance or whatever we are hosting this application on. @marshallj @ejaxon I was hoping one of you could add some input to this issue.

File: src/components/Contact.jsx [hunk 2]
```diff
@@ -1,21 +1,152 @@
...
+    const opts = {
+      method: 'POST',
+      body: JSON.stringify(this.state),
+      headers: new Headers({
+        'Content-Type': 'application/json',
+      }),
+    };
+
+    // this url needs to be wherever the server is getting hosted. Obviously localhost:3001 will not work on a production build.
+    const url = process.env.API_URL || 'http://localhost:3001';
```
Comment (BlakeGuilloud) [inline src/components/Contact.jsx]: This is the big question I have. I need to know what to set this URL to in order to properly make an API request. An option would be to set this as a process.env variable and point it to the EC2 instance or whatever we are hosting this application on. @marshallj @ejaxon I was hoping one of you could add some input to this issue.

Comment (ejaxon) [reply src/components/Contact.jsx]: @BlakeGuilloud The domain for the site will be ncreentryhub.org, although we don't have the web server there running ATM. I just set the DNS to point to the AWS instance, but IP is 52.4.230.168. 

I like the idea of putting the domain in the .env.

BTW, I assume this won't hurt, but there will actually be multiple URLs used to access the site. ncreentryhub.org is the main one, but individual counties may access via separate URLs, e.g., we'll eventually point buncombereentryhub.org there and the URL will be used to determine the correct locale.

Does that work?

**Comment by ejaxon** [comment]
Also, I'd rather go ahead and just have the 6 environment variables rather than hardcoding.

**Comment by BlakeGuilloud** [comment]
Sweet. Thanks for the feedback. I'll push up some changes after work tonight.

**Comment by BlakeGuilloud** [comment]
Alright @ejaxon looks like this should be about ready to go. I have added a few more env variables all of which are demonstrated in `sample.env`. All developers will have to create a `.env` file locally (it is gitignored as it contains sensitive info.) and fill it with appropriate values:

```
AWS_KEY = 123
AWS_SECRET = 456
AWS_DEFAULT_REGION = us-east-1
SES_EMAIL = test@test.com
RECEPIENT_EMAILS = abc@gmail.com,def@hotmail.com
REACT_APP_API_URL = http://localhost:3001
```

If you want to try the feature out yourself, you can verify your email address with the SES service and replace the RECEIPIENT_EMAILS variable to include your newly verified email address. Once this goes in, we will need to make sure we included all of these variables in the node process somehow. I'm not much of a devops guy, so I might not be of much help with that. I'm familiar w/ jenkins and circleci but we obviously don't have anything like that setup for this project.

Let me know if you have any questions, but I'm going to go ahead and remove the do not merge tag and hopefully we can get this guy in at some point this week.

Again- let me know if you have any questions or if there is anything I can help with. I said I'm not much of a devops guy, but I've been trying to expose myself to all things AWS- so I'm happy to lend a hand there as well.

**Comment by ejaxon** [comment]
It'll take me a couple days to get to a detailed review, but the main question is whether anything *besides* the contact page breaks if developers don't include a .env file? Having a functioning contact page is only relevant to developers actually working on that. Though we definitely need to add documentation of this to the README.

In any case, I'll look by Friday at latest.

**Comment by BlakeGuilloud** [comment]
@ejaxon To answer you question. No. Nothing will break for any developer that does not have a .env file, but they will not be able to submit the contact form - I would be happy to write up some documentation for the ReadMe

**Comment by ejaxon** [comment]
@BlakeGuilloud I assumed as much, but wanted to explicitly check. I'm going to go ahead and merge. Thanks for this! 

I'm going to create a separate issue for updating the readme and will tag you there.

**Pull request merged**